tests/sample_qc/test_de_novo.py

```diff
@@ -0,0 +1,230 @@
+"""Test suite for de novo mutation functions."""
+
+import hail as hl
+import pytest
+
+from gnomad.sample_qc.relatedness import (
+    calculate_de_novo_post_prob,
+    call_de_novo,
+    get_de_novo_expr,
+    get_freq_prior,
+    transform_pl_to_pp,
+)
+from gnomad.utils.annotations import get_copy_state_by_sex
+
+
+class TestDeNovoMutation:
+    """Test suite for de novo mutation functions."""
+
+    loci: dict[str, hl.expr.LocusExpression]
+
+    @classmethod
+    def setup_class(cls):
+        """Set up common test data for all tests."""
+        cls.locus_expr = hl.locus("1", 123456)
+        cls.alleles_expr = hl.literal(["A", "T"])
+        cls.freq_prior_expr = hl.literal(0.01)
+        cls.is_xx_expr = hl.literal(False)
+
+        # Mock Genotype Likelihoods (PL)
+        cls.proband_pl = hl.literal([0, 10, 100])
+        cls.father_pl = hl.literal([0, 100, 100])
+        cls.mother_pl = hl.literal([0, 100, 100])
+
+        # Mock Genotype Calls
+        cls.proband_expr = hl.struct(
+            GT=hl.call(0, 1), DP=10, GQ=30, AD=[3, 7], PL=cls.proband_pl
+        )
+        cls.father_expr = hl.struct(
+            GT=hl.call(0, 0), DP=12, GQ=40, AD=[12, 0], PL=cls.father_pl
+        )
+        cls.mother_expr = hl.struct(
+            GT=hl.call(0, 0), DP=15, GQ=50, AD=[15, 0], PL=cls.mother_pl
+        )
+
+        cls.loci = {
+            "autosomal": hl.locus("chr1", 100000, reference_genome="GRCh38"),
+            # PAR regions (always diploid)
+            "par1": hl.locus("chrX", 2781479, reference_genome="GRCh38"),  # PAR1 start
+            "par2": hl.locus("chrX", 155701383, reference_genome="GRCh38"),  # PAR2 end
+            # X non-PAR (diploid for XX, hemizygous for XY)
+            "x_nonpar": hl.locus("chrX", 3000000, reference_genome="GRCh38"),
+            # Y non-PAR (hemizygous for XY)
+            "y_nonpar": hl.locus("chrY", 10000000, reference_genome="GRCh38"),
+        }
+
+    @pytest.mark.parametrize(
+        "freq_prior, min_pop_prior, expect_error, expected",
+        [
+            (0.05, 100 / 3e7, False, 0.05),
+            (-0.01, 100 / 3e7, True, None),
+            (1.2, 100 / 3e7, True, None),
+            (hl.missing(hl.tfloat64), 100 / 3e7, False, 100 / 3e7),
+        ],
+    )
+    def test_get_freq_prior(
+        self, freq_prior, min_pop_prior, expect_error, expected
+    ) -> None:
+        """
+        Test frequency prior computation.
+
+        :param freq_prior: Frequency prior value.
+        :param min_pop_prior: Minimum population prior.
+        :param expect_error: Whether an error is expected.
+        :param expected: Expected frequency prior.
+        :return: None.
+        """
+        if expect_error:
+            with pytest.raises(
+                hl.utils.java.HailUserError,
+                match="de_novo: expect 0 <= freq_prior_expr <= 1",
```

Comment:
naive question, does this `match` work with the error thrown by the actual function since that error message will include the user-provided freq prior? `de_novo: expect 0 <= freq_prior_expr <= 1, found %.3e`